Null Pointer while validating with $ref field

### Version
4.1.1
### Context

While updating a project to use vertx-json-schema to validate JSON schemas based on TS 3GPP documents read from files, NullPointerExceptions and ConcurrentModificationExceptions are encountered. These exceptions are inconsistent, and always happen in Vertx code. Additionally, sometimes, the "trying to execute sync validation in async state" exception is encountered when calling validateAsync.

### Do you have a reproducer?

No

### Steps to reproduce

1. Write schemas that use $refs. The project has many, many schemas that reference other schemas.
2. Use validateAsync to validate many of the schemas. The project contains test classes for 20-30 of them, each test class containing test methods that assert that validation works as expected.

### Extra


`validation.JsonValidationException: ValidationException{message='additionalProperties schema should match', keyword='additionalProperties', input={"additionalProperties":{"supiRangeList":[{"start":"111","end":"999","pattern":"pattern"}],"gpsiRangeList":[{"start":"123","end":"789","pattern":"pattern-value"}],"plmnRangeList":[{"start":"11122","end":"11133","pattern":"pattern-value"}],"groupId":"123456789","primaryChfInstance":"6f134298-939a-47d6-9566-fd0030517ac2","secondaryChfInstance":"66ac6eea-db7c-44e4-a8e8-2a5d6e5184ef"}}, schema=io.vertx.json.schema.common.SchemaImpl@7e7a6f81, inputScope=#}
	at validation.JsonValidationUtils.validateObject(JsonValidationUtils.java:74)
	at validation.ts3gpp.extensions.ChfInfoListValidator.validate(ChfInfoListValidator.java:41)
	at validation.ts3gpp.extensions.ChfInfoListNGTest.validate(ChfInfoListNGTest.java:53)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:135)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeMulti(TestNGDirectoryTestSuite.java:193)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:94)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: ValidationException{message='additionalProperties schema should match', keyword='additionalProperties', input={"additionalProperties":{"supiRangeList":[{"start":"111","end":"999","pattern":"pattern"}],"gpsiRangeList":[{"start":"123","end":"789","pattern":"pattern-value"}],"plmnRangeList":[{"start":"11122","end":"11133","pattern":"pattern-value"}],"groupId":"123456789","primaryChfInstance":"6f134298-939a-47d6-9566-fd0030517ac2","secondaryChfInstance":"66ac6eea-db7c-44e4-a8e8-2a5d6e5184ef"}}, schema=io.vertx.json.schema.common.SchemaImpl@7e7a6f81, inputScope=#}
	at io.vertx.json.schema.ValidationException.create(ValidationException.java:71)
	at io.vertx.json.schema.common.PropertiesValidatorFactory.fillAdditionalPropertyException(PropertiesValidatorFactory.java:111)
	at io.vertx.json.schema.common.PropertiesValidatorFactory.access$200(PropertiesValidatorFactory.java:32)
	at io.vertx.json.schema.common.PropertiesValidatorFactory$PropertiesValidator.lambda$validateAsync$0(PropertiesValidatorFactory.java:206)
	at io.vertx.core.impl.future.ComposeTransformation.onFailure(ComposeTransformation.java:50)
	at io.vertx.core.impl.future.FutureBase.emitFailure(FutureBase.java:78)
	at io.vertx.core.impl.future.FutureImpl.addListener(FutureImpl.java:160)
	at io.vertx.core.impl.future.FutureBase.compose(FutureBase.java:104)
	at io.vertx.core.Future.recover(Future.java:210)
	at io.vertx.json.schema.common.PropertiesValidatorFactory$PropertiesValidator.validateAsync(PropertiesValidatorFactory.java:206)
	at io.vertx.json.schema.common.SchemaImpl.runAsyncValidators(SchemaImpl.java:173)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:121)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:48)
	atvalidation.JsonValidationUtils.validateObject(JsonValidationUtils.java:60)
	... 32 more
Caused by: java.lang.NullPointerException
	at io.vertx.json.schema.common.SchemaImpl.runAsyncValidators(SchemaImpl.java:171)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:121)
	at io.vertx.json.schema.common.RefSchema.lambda$validateAsync$1(RefSchema.java:79)
	at io.vertx.core.impl.future.ComposeTransformation.onSuccess(ComposeTransformation.java:38)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:61)
	at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:82)
	at io.vertx.core.impl.future.FutureBase.compose(FutureBase.java:104)
	at io.vertx.core.impl.future.SucceededFuture.compose(SucceededFuture.java:27)
	at io.vertx.json.schema.common.RefSchema.validateAsync(RefSchema.java:65)
	at io.vertx.json.schema.common.PropertiesValidatorFactory$PropertiesValidator.validateAsync(PropertiesValidatorFactory.java:205)`

schema below

`Caused by: io.vertx.json.schema.NoSyncValidationException: Trying to execute validateSync() for a Validator in asynchronous state
	at io.vertx.json.schema.common.BaseMutableStateValidator.checkSync(BaseMutableStateValidator.java:59)
	at io.vertx.json.schema.common.SchemaImpl.validateSync(SchemaImpl.java:126)
	at io.vertx.json.schema.common.RefSchema.validateSync(RefSchema.java:104)
	at io.vertx.json.schema.common.ItemsValidatorFactory$ItemsValidator.validateSync(ItemsValidatorFactory.java:49)
	at io.vertx.json.schema.common.SchemaImpl.runSyncValidator(SchemaImpl.java:193)
	at io.vertx.json.schema.common.SchemaImpl.validateSync(SchemaImpl.java:128)
	at io.vertx.json.schema.common.PropertiesValidatorFactory$PropertiesValidator.validateAsync(PropertiesValidatorFactory.java:166)
	at io.vertx.json.schema.common.SchemaImpl.runAsyncValidators(SchemaImpl.java:173)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:121)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:48)
	at validation.JsonValidationUtils.validateObject(JsonValidationUtils.java:60)`

schema:

`{
    "description": "TS29510_Nnrf_NFManagement.yaml",
    "javaType": "ts3gpp.ChfInfo",
    "type": "object",
    "properties": {
        "supiRangeList": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "supi-range.json"
            }
        },
        "gpsiRangeList": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "identity-range.json"
            }
        },
        "plmnRangeList": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "plmn-range.json"
            }
        },
        "groupId": {
            "type": "string"
        },
        "primaryChfInstance": {
            "type": "string",
            "pattern": "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
        },
        "secondaryChfInstance": {
            "type": "string",
            "pattern": "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
        }
    },
    "required": [
        "primaryChfInstance",
        "secondaryChfInstance"
    ]
}`

---------------------------------

`java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.forEach(HashMap.java:1339)
	at io.vertx.json.schema.common.SchemaRouterImpl.lambda$getScopeParentAliases$10(SchemaRouterImpl.java:215)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at io.vertx.json.schema.common.SchemaRouterImpl.getScopeParentAliases(SchemaRouterImpl.java:215)
	at io.vertx.json.schema.common.SchemaRouterImpl.resolveAbsoluteUriAlternatives(SchemaRouterImpl.java:227)
	at io.vertx.json.schema.common.SchemaRouterImpl.resolveParentNode(SchemaRouterImpl.java:245)
	at io.vertx.json.schema.common.SchemaRouterImpl.resolveCachedSchema(SchemaRouterImpl.java:65)
	at io.vertx.json.schema.common.SchemaRouterImpl.resolveRef(SchemaRouterImpl.java:105)
	at io.vertx.json.schema.common.RefSchema.validateAsync(RefSchema.java:64)
	at io.vertx.json.schema.common.ItemsValidatorFactory$ItemsValidator.validateAsync(ItemsValidatorFactory.java:63)
	at io.vertx.json.schema.common.SchemaImpl.runAsyncValidators(SchemaImpl.java:173)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:121)
	at io.vertx.json.schema.common.PropertiesValidatorFactory$PropertiesValidator.validateAsync(PropertiesValidatorFactory.java:171)
	at io.vertx.json.schema.common.SchemaImpl.runAsyncValidators(SchemaImpl.java:173)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:121)
	at io.vertx.json.schema.common.SchemaImpl.validateAsync(SchemaImpl.java:48)`

schema:
`{
    "description": "TS29510_Nnrf_NFManagement.yaml",
    "javaType": "ts3gpp.LmfInfo",
    "type": "object",
    "properties": {
        "externalClientType": {
            "type": "array",
            "default": null,
            "items": {
                "$ref": "enums/external-client-type.json",
                "existingJavaType": "ts3gpp.enums.ExternalClientType"
            }
        },
        "lmfId": {
            "type": "string"
        },
        "servingAccessTypes": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "enums/access-type.json",
                "existingJavaType": "ts3gpp.enums.AccessType"
            }
        },
        "servingAnNodeTypes": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "enums/an-node-type.json",
                "existingJavaType": "ts3gpp.enums.AnNodeType"
            }
        },
        "servingRatTypes": {
            "type": "array",
            "default": null,
            "minItems": 1,
            "items": {
                "$ref": "enums/rat-type.json",
                "existingJavaType": "ts3gpp.enums.RatType"
            }
        }
    }
}`

--- a/src/test/java/io/vertx/json/schema/ValidationTest.java
+++ b/src/test/java/io/vertx/json/schema/ValidationTest.java
@@ -1,13 +1,13 @@
 package io.vertx.json.schema;
 
 import io.vertx.core.Vertx;
-import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 import io.vertx.junit5.VertxExtension;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
 
 @ExtendWith(VertxExtension.class)
 class ValidationTest {
@@ -70,5 +70,39 @@ class ValidationTest {
 
     assertThat(res.getValid()).isTrue();
     assertThat(res.getErrors()).isNull();
+  }
+
+  @Test
+  public void testComplexSchema() throws JsonSchemaValidationException {
+    final Validator validator = Validator.create(
+      JsonSchema.of(new JsonObject("{ \"description\": \"TS29510_Nnrf_NFManagement.yaml\", \"javaType\": \"ts3gpp.ChfInfo\", \"type\": \"object\", \"properties\": { \"supiRangeList\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"supi-range.json\" } }, \"gpsiRangeList\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"identity-range.json\" } }, \"plmnRangeList\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"plmn-range.json\" } }, \"groupId\": { \"type\": \"string\" }, \"primaryChfInstance\": { \"type\": \"string\", \"pattern\": \"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$\" }, \"secondaryChfInstance\": { \"type\": \"string\", \"pattern\": \"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$\" } }, \"required\": [ \"primaryChfInstance\", \"secondaryChfInstance\" ] }")),
+      new JsonSchemaOptions()
+        .setBaseUri("https://vertx.io")
+        .setDraft(Draft.DRAFT7)
+        .setOutputFormat(OutputFormat.Basic));
+
+    final OutputUnit res = validator.validate(new JsonObject("{\"additionalProperties\":{\"supiRangeList\":[{\"start\":\"111\",\"end\":\"999\",\"pattern\":\"pattern\"}],\"gpsiRangeList\":[{\"start\":\"123\",\"end\":\"789\",\"pattern\":\"pattern-value\"}],\"plmnRangeList\":[{\"start\":\"11122\",\"end\":\"11133\",\"pattern\":\"pattern-value\"}],\"groupId\":\"123456789\",\"primaryChfInstance\":\"6f134298-939a-47d6-9566-fd0030517ac2\",\"secondaryChfInstance\":\"66ac6eea-db7c-44e4-a8e8-2a5d6e5184ef\"}}"));
+
+    try {
+      res.checkValidity();
+      fail("Should have thrown an exception");
+    } catch (JsonSchemaValidationException e) {
+      // OK
+    }
+  }
+
+  @Test
+  public void testComplexSchema2() throws JsonSchemaValidationException {
+    final Validator validator = Validator.create(
+      JsonSchema.of(new JsonObject("{ \"description\": \"TS29510_Nnrf_NFManagement.yaml\", \"javaType\": \"ts3gpp.LmfInfo\", \"type\": \"object\", \"properties\": { \"externalClientType\": { \"type\": \"array\", \"default\": null, \"items\": { \"$ref\": \"enums/external-client-type.json\", \"existingJavaType\": \"ts3gpp.enums.ExternalClientType\" } }, \"lmfId\": { \"type\": \"string\" }, \"servingAccessTypes\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"enums/access-type.json\", \"existingJavaType\": \"ts3gpp.enums.AccessType\" } }, \"servingAnNodeTypes\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"enums/an-node-type.json\", \"existingJavaType\": \"ts3gpp.enums.AnNodeType\" } }, \"servingRatTypes\": { \"type\": \"array\", \"default\": null, \"minItems\": 1, \"items\": { \"$ref\": \"enums/rat-type.json\", \"existingJavaType\": \"ts3gpp.enums.RatType\" } } } }")),
+      new JsonSchemaOptions()
+        .setBaseUri("https://vertx.io")
+        .setDraft(Draft.DRAFT7)
+        .setOutputFormat(OutputFormat.Basic));
+
+    final OutputUnit res = validator.validate(new JsonObject("{\"additionalProperties\":{\"supiRangeList\":[{\"start\":\"111\",\"end\":\"999\",\"pattern\":\"pattern\"}],\"gpsiRangeList\":[{\"start\":\"123\",\"end\":\"789\",\"pattern\":\"pattern-value\"}],\"plmnRangeList\":[{\"start\":\"11122\",\"end\":\"11133\",\"pattern\":\"pattern-value\"}],\"groupId\":\"123456789\",\"primaryChfInstance\":\"6f134298-939a-47d6-9566-fd0030517ac2\",\"secondaryChfInstance\":\"66ac6eea-db7c-44e4-a8e8-2a5d6e5184ef\"}}"));
+
+    // Should be fine!
+    res.checkValidity();
   }
 }

